hpx/parallel/datapar/loop.hpp

```diff
@@ -0,0 +1,615 @@
+    HPX_HOST_DEVICE HPX_FORCEINLINE
+    auto loop_step(parallel::v1::datapar_task_execution_policy, F && f,
+            Iter1& it1, Iter2& it2)
+    ->  decltype(
+            detail::datapar_loop_step2<Iter1, Iter2>::callv(
+                std::forward<F>(f), it1, it2)
+            )
+    {
+        return detail::datapar_loop_step2<Iter1, Iter2>::callv(
+            std::forward<F>(f), it1, it2);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        template <typename Iter, typename Enable = void>
+        struct loop_optimization
+        {
+            template <typename Iter1>
+            static bool call(Iter1 const& first1, Iter1 const& last1)
+            {
+                return false;
+            }
+        };
+
+        template <typename Iter>
+        struct loop_optimization<Iter,
+            typename std::enable_if<
+                iterator_datapar_compatible<Iter>::value
+            >::type>
+        {
+            template <typename Iter1>
+            static bool call(Iter1 const& first1, Iter1 const& last1)
+            {
+                typedef typename std::iterator_traits<Iter1>::value_type
+                    value_type;
+                typedef Vc::Vector<value_type> V;
+
+                return V::Size <= std::distance(first1, last1);
+            }
+        };
+    }
+
+    template <typename Iter>
+    HPX_HOST_DEVICE HPX_FORCEINLINE
+    bool loop_optimization(parallel::v1::datapar_execution_policy,
+        Iter const& first1, Iter const& last1)
+    {
+        return detail::loop_optimization<Iter>::call(first1, last1);
+    }
+
+    template <typename Iter>
+    HPX_HOST_DEVICE HPX_FORCEINLINE
+    bool loop_optimization(parallel::v1::datapar_task_execution_policy,
+        Iter const& first1, Iter const& last1)
+    {
+        return detail::loop_optimization<Iter>::call(first1, last1);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        ///////////////////////////////////////////////////////////////////////
+        template <typename T, typename Abi>
+        HPX_HOST_DEVICE HPX_FORCEINLINE
+        std::size_t count_bits(Vc::Mask<T, Abi> const& mask)
+        {
+            return mask.count();
+        }
+
+        template <typename F, typename Vector>
+        HPX_HOST_DEVICE HPX_FORCEINLINE
+        typename hpx::util::decay<Vector>::type::EntryType
+        accumulate_values(parallel::v1::datapar_execution_policy,
+            F && f, Vector && value)
+        {
+            typedef typename hpx::util::decay<Vector>::type vector_type;
+            typedef typename vector_type::EntryType entry_type;
+
+            entry_type accum = value[0];
+            for(size_t i = 1; i != value.size(); ++i)
+            {
+                accum = f(accum, entry_type(value[i]));
+            }
+            return accum;
```

Comment:
You're right. I was planning to do that but forgot. `Vc::Vector::sum` does exactly what we need.